File: WebContent/components/signup/signup.js
ratorApp.controller('signupController', function($scope, $http, $location, requestService) {
	
	// Sign up
	$scope.signup = function(){
		
		$scope.result = "Waiting for response";
		
		var signupReq = requestService.buildRequest(
				"Signup", 
				{
					'username': $scope.username,
					'password': $scope.password,
					'retype': $scope.retype,
					'playername': $scope.playername
				}
			);
		
		var minlength = function(s){
			if(s.length > 7){
				return true;
			}
			return false;
		}
		
		var match = function(a, b){
			if (a === b){
				return true;
			}
			return false;
		}
				
		if( minlength($scope.password) ){
			
			if( match($scope.password, $scope.retype) ){
	
				$http(signupReq).then(function(response){
					
						$scope.result = response.data.result;
					
						if (response.data.result == "Success"){
							$location.url('/');				
						}
					}, 
					function(){
						$scope.result = 'Request failure';
				});		
			} else {
				$scope.result = 'Passwords does not match';
			}
			
		} else {
			$scope.result = 'Passwords is to short, minimum length is 8 characters';
		}
	};
});

File: src/magirator/servlets/Signup.java
package magirator.servlets;

import java.io.IOException;
import javax.servlet.ServletException;
import javax.servlet.annotation.WebServlet;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import com.google.gson.JsonObject;

import magirator.logic.LoginCredentials;
import magirator.model.neo4j.Players;
import magirator.model.neo4j.Users;
import magirator.support.Error;
import magirator.support.Json;
import magirator.support.Constants;
import magirator.view.PlayerName;

/**
 * Servlet implementation class Signup
 */
@WebServlet("/Signup")
public class Signup extends HttpServlet {
	private static final long serialVersionUID = 1L;

	/**
	 * @see HttpServlet#doPost(HttpServletRequest request, HttpServletResponse response)
	 */
	protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
		
		getServletContext().log("-- Signup --");
		
		JsonObject result = new JsonObject();
		result.addProperty(Constants.result, "Something went wrong during singup, please try again");
		
		try {
			
			JsonObject signupRequest = Json.parseRequestData(request);
			LoginCredentials requestedUser = new LoginCredentials(signupRequest);
			
			result.addProperty(Constants.result, "Password length is to short, please choose a password at least 8 characters long");
			
			if(requestedUser.getPassword().length() > 7){
			
				result.addProperty(Constants.result, "Passwords does not match, please check and try again");
				
				if(requestedUser.getPassword().equals(requestedUser.getRetype())){
	
					result.addProperty(Constants.result, "Username already in use, please try another one");
					
					if (Users.nameIsAvailable(requestedUser)){
						
						PlayerName requestedPlayer = new PlayerName(signupRequest);
		
						result.addProperty(Constants.result, "Playername already in use, please try another one");
						
						boolean playerAvailable = Players.checkIfAvailable(requestedPlayer);
						
						if (playerAvailable){
							
							result.addProperty(Constants.result, "This user and player name really should be ok, please try again");
							
							requestedUser.encryptPassword();
							
							boolean signupSuccessful = Users.signup(requestedUser, requestedPlayer);
							
							if (signupSuccessful){
								
								result.addProperty(Constants.result, Constants.success);
							}					
						}
					}
				}
			}
		} catch (Exception e) {
			result.addProperty(Constants.result, Error.printStackTrace(e));
		}

		response.setContentType("application/json");
		response.getWriter().write(result.toString());
		
		getServletContext().log("-- Signup Done --");
	}

}
